keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,409 @@
+   ready on the Node: then Kube-proxy will indicate to the load balancer that it
+   should send it traffic. This will not be the case anymore once this change is
+   implemented. Clients relying on this behavior will probably see their
+   connections fail.
+
+#### Mitigations
+
+1. If Kube-proxy starts failing when reading from the API server, it should just
+   assume that the last state seen continues. For Kube-proxy to be made aware of
+   this, it needs to invoke
+   `serviceInformer.Informer().SetWatchErrorHandler(DefaultWatchErrorHandler)`
+   when initializing its informers. Any errors observed by client-go when
+   watching from the API server will be reported on `DefaultWatchErrorHandler`.
+
+2. Metrics should inform on Kube-proxy health and include information about its
+   `healthz` state, this can then be used to correlate to networking metrics
+   surrounding new/established connections on the node. Ex: a failing `healthz`
+   should correlate to a total drop in the count of new connections and with a
+   zero-or-negative rate of established connections. E2E tests should also be
+   designed with this specific goal in mind, i.e: validating the impact of a
+   failing kube-proxy on ingress connectivity. Kube-proxy currently has a lot of
+   metrics regarding how its health is doing, but nothing no sole red/green
+   indicator of what the end result of its health is. One such metric could be
+   `proxy_health_check_200_count` / `proxy_health_check_503_count`
+
+3. The feature could be disabled for user who is dependent upon such behavior by
+   means of flipping the feature flag to off.
+
+## Design Details
+
+1. Implement change in Kube-proxy which allows the healthcheck server access to
+   the `healthz` state. Implement change required to be notified by client-go on
+   watch errors from the API server.
+
+2. Implement change in Kube-proxy which will react to changes on the Node object
+   and once the taint `ToBeDeletedByClusterAutoscaler` is placed on the Node
+   object: start failing it's `healthz` state.
+
+3. Write document to be published at: https://kubernetes.io/docs/concepts/ which
+   details: a) how determining node/instance health can best be done for
+   Kubernetes clusters b) how Kube-proxy will do it once the changes proposed in
+   this KEP are merged c) what some pitfalls with other methods might be.
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+- Add E2E tests for connectivity to services on terminating nodes and
+  validate graceful termination of TCP connections.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Preferably E2E tests coded before any feature implementation is made which
```

Comment:
why preferably? the feature has to merge with the e2e tests or we can't know if it works

Reply:
It's more because of the sensitivity of these tests and the feasibility (which I don't know at this point). We need a stable CI signal for tests that need to break nodes (specifically have kube-proxy report unhealthy) and inspect the state of TCP connections passing through said node, sounds like flake heaven.

Reply:
we have tests with disruption that are not flaky https://testgrid.k8s.io/sig-network-gce#gci-gce-ingress&include-filter-by-regex=disruption

the e2e should describe user behavior, the user should should see connection drops, no need to get into the nodes to check for TCP connections

Reply:
Oh, that's cool! I am updating the PR now, I will mention this.